File: src/generator.js
/**
 * Public generator entry point. generateL1Voronoi composes the pipeline
 * stages (preprocess -> sort/init -> divide & conquer -> polygonize). The
 * naive brute-force generator lives in tests/ as an oracle.
 */

import {recursiveSplit} from './divideConquer.js';
import {polygonizeSite} from './polygonizer.js';
import {toCell} from './cell.js';
import {cleanData} from './preprocess.js';
import {createL1Metric} from './l1Metric.js';
import {createSite} from './bisector.js';

function compareByXY(a, b){
    if(a[0] !== b[0]){
        return a[0] - b[0];
    }
    return a[1] - b[1];
}

/**
 * Generate an L1 Voronoi diagram using Lee & Wong's algorithm.
 *
 * @param {array} sitePoints
 * @param {number} width
 * @param {number} height
 * @param {boolean} nudgeData
 * @returns {Array<Site>}
 */
export function generateL1Voronoi(sitePoints, width, height, nudgeData = true){

    let workingPoints = sitePoints;
    if(nudgeData){
        // nudge a copy so the caller's input is never modified
        workingPoints = cleanData(sitePoints.map(p => p.slice()));
    }

    let sites = workingPoints.slice().sort(compareByXY).map(createSite);

    const metric = createL1Metric(width, height);
    const graph = recursiveSplit(sites, metric);

    const cells = graph.map(site => polygonizeSite(site, metric));

    return cells.map(toCell);
}

File: py_ai/generator.py
"""
Public generator entry points. generate_l1_voronoi composes the pipeline
stages (preprocess -> sort/init -> divide & conquer -> polygonize); the
exported functions are the library facade.
"""

from divide_conquer import recursive_split
from polygonizer import polygonize_site
from cell import to_cell
from preprocess import clean_data
from l1_metric import create_l1_metric
from bisector import create_site


def generate_l1_voronoi(site_points, width, height, nudge_data=True):
    """Generate an L1 Voronoi diagram"""
    if nudge_data:
        site_points = clean_data(site_points[:])

    sorted_points = sorted(site_points, key=lambda p: (p[0], p[1]))
    sites = [create_site(e) for e in sorted_points]

    metric = create_l1_metric(width, height)
    graph = recursive_split(sites, metric)

    polygonized = [polygonize_site(site, metric) for site in graph]

    return [to_cell(site) for site in polygonized]
